controllers/argocd/prometheus.go

```diff
@@ -247,3 +248,150 @@ func (r *ReconcileArgoCD) reconcileServerMetricsServiceMonitor(cr *argoprojv1a1.
+					},
+					For: "1m",
+					Labels: map[string]string{
+						"severity": "critical",
+					},
+				},
+				{
+					Alert: "ApplicationSetControllerNotReady",
+					Annotations: map[string]string{
+						"message": fmt.Sprintf("applicationSet controller deployment for Argo CD instance in namespace %s is not running", cr.Namespace),
+					},
+					Expr: intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: fmt.Sprintf("kube_deployment_status_replicas{deployment=\"%s\", namespace=\"%s\"} != kube_deployment_status_replicas_ready{deployment=\"%s\", namespace=\"%s\"} ", fmt.Sprintf(cr.Name+"-applicationset-controller"), cr.Namespace, fmt.Sprintf(cr.Name+"-applicationset-controller"), cr.Namespace),
+					},
+					For: "10m",
+					Labels: map[string]string{
+						"severity": "warning",
+					},
+				},
+				{
+					Alert: "DexNotReady",
+					Annotations: map[string]string{
+						"message": fmt.Sprintf("dex deployment for Argo CD instance in namespace %s is not running", cr.Namespace),
+					},
+					Expr: intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: fmt.Sprintf("kube_deployment_status_replicas{deployment=\"%s\", namespace=\"%s\"} != kube_deployment_status_replicas_ready{deployment=\"%s\", namespace=\"%s\"} ", fmt.Sprintf(cr.Name+"-dex-server"), cr.Namespace, fmt.Sprintf(cr.Name+"-dex-server"), cr.Namespace),
+					},
+					For: "10m",
+					Labels: map[string]string{
+						"severity": "warning",
+					},
+				},
+				{
+					Alert: "NotificationsControllerNotReady",
+					Annotations: map[string]string{
+						"message": fmt.Sprintf("notifications controller deployment for Argo CD instance in namespace %s is not running", cr.Namespace),
+					},
+					Expr: intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: fmt.Sprintf("kube_deployment_status_replicas{deployment=\"%s\", namespace=\"%s\"} != kube_deployment_status_replicas_ready{deployment=\"%s\", namespace=\"%s\"} ", fmt.Sprintf(cr.Name+"-notifications-controller"), cr.Namespace, fmt.Sprintf(cr.Name+"-notifications-controller"), cr.Namespace),
+					},
+					For: "10m",
+					Labels: map[string]string{
+						"severity": "warning",
+					},
+				},
+				{
+					Alert: "RedisNotReady",
+					Annotations: map[string]string{
+						"message": fmt.Sprintf("redis deployment for Argo CD instance in namespace %s is not running", cr.Namespace),
+					},
+					Expr: intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: fmt.Sprintf("kube_deployment_status_replicas{deployment=\"%s\", namespace=\"%s\"} != kube_deployment_status_replicas_ready{deployment=\"%s\", namespace=\"%s\"} ", fmt.Sprintf(cr.Name+"-redis"), cr.Namespace, fmt.Sprintf(cr.Name+"-redis"), cr.Namespace),
+					},
+					For: "10m",
+					Labels: map[string]string{
+						"severity": "warning",
+					},
+				},
+			},
+		},
+	}
+	promRule.Spec.Groups = ruleGroups
+
+	if err := controllerutil.SetControllerReference(cr, promRule, r.Scheme); err != nil {
```

Comment:
Hmmm
this seems tricky
This part of the code creates a new promrule with the cr name/namespace by default
if i pass a 2nd cr into this function, it'll just create a 2nd rule and work normally
if the 2nd cr is in the same namespace then the rule will already exist and the code will return nil

I'm not sure i can invoke just this part of the function by passing in a different cr but the same promRule

Reply:
Indeed this is tricky... Hmmm, can you try to mock it?

```go
cr := &runtime.Object{}
promRule := &monitoringv1.PrometheusRule{}

// Test error case
controllerutil.SetControllerReference = func(owner, obj runtime.Object, scheme *runtime.Scheme) error {
    return errors.NewBadRequest("error")
}

// Call it and error should not be nil
err = controllerutil.SetControllerReference(cr, promRule, r.Scheme)
```

If this doesn't work, don't waste more time here. The important stuff are tested. This is just me being nitty-picky/gold-plating.

Reply:
calling `setControllerReference` explicity won't test the function code though since ideally we would want that to be invoked through the function we are testing 

to be honest, I'm not sure how much value there would be in testing this scenario since this is a pretty common thing we do with every resource we create/reconcile in the operator